xds/src/main/java/io/grpc/xds/XdsNameResolver.java

```diff
@@ -89,49 +108,82 @@ public void start(final Listener2 listener) {
+        String serviceConfig = "{\n"
+            + "  \"loadBalancingConfig\": [\n"
+            + "    {\n"
+            + "      \"experimental_cds\": {\n"
+            + "        \"cluster\": \"" + update.getClusterName() + "\"\n"
+            + "      }\n"
+            + "    }\n"
+            + "  ]\n"
+            + "}";
+        Map<String, ?> config;
+        try {
+          config = (Map<String, ?>) JsonParser.parse(serviceConfig);
+        } catch (IOException e) {
+          listener.onError(
+              Status.UNKNOWN.withDescription("Invalid service config").withCause(e));
+          return;
+        }
+        Attributes attrs =
+            Attributes.newBuilder()
+                .set(GrpcAttributes.NAME_RESOLVER_SERVICE_CONFIG, config)
+                .set(XdsAttributes.XDS_CLIENT_REF, xdsClientPool)
+                .build();
+        ResolutionResult result =
+            ResolutionResult.newBuilder()
+                .setAddresses(ImmutableList.<EquivalentAddressGroup>of())
+                .setAttributes(attrs)
+                .build();
+        listener.onResult(result);
+      }
 
-    String serviceConfig = "{\n"
-        + "  \"loadBalancingConfig\": [\n"
-        + "    {\n"
-        + "      \"xds_experimental\": {\n"
-        + "        \"childPolicy\": [ {\"round_robin\": {} } ]\n"
-        + "      }\n"
-        + "    }"
-        + "  ]\n"
-        + "}";
-    Map<String, ?> config;
-    try {
-      config = (Map<String, ?>) JsonParser.parse(serviceConfig);
-    } catch (IOException e) {
-      listener.onError(
-          Status.UNKNOWN.withDescription("Invalid service config").withCause(e));
-      throw new AssertionError("Invalid service config");
-    }
-    Attributes attrs =
-        Attributes.newBuilder()
-            .set(GrpcAttributes.NAME_RESOLVER_SERVICE_CONFIG, config)
-            .set(XDS_NODE, bootstrapInfo.getNode())
-            .set(XDS_CHANNEL_CREDS_LIST, serverInfo.getChannelCredentials())
-            .build();
-    ResolutionResult result =
-        ResolutionResult.newBuilder()
-            .setAddresses(Collections.<EquivalentAddressGroup>emptyList())
-            .setAttributes(attrs)
-            .build();
-    listener.onResult(result);
+      @Override
+      public void onError(Status error) {
+        // In order to distinguish between IO error and resource not found, which trigger
+        // different handling, return an empty resolution result to channel for resource not
+        // found.
+        // TODO(chengyuanzhang): Returning an empty resolution result based on status code is
+        //  a temporary solution. More design discussion needs to be done.
+        if (error.getCode().equals(Code.NOT_FOUND)) {
+          listener.onResult(ResolutionResult.newBuilder().build());
+        }
+        listener.onError(error);
+      }
+    });
   }
 
   @Override
   public void shutdown() {
+    if (xdsClient != null) {
+      xdsClientPool.returnObject(xdsClient);
```

Comment:
`NameResolver.shutdown()` spec didn't say this should be called no more than once.

To be safe set it to null `xdsClient = xdsClientPool.returnObject(xdsClient);`

Reply:
I think I will also update this pattern to CDS/EDS balancers, even `balancer.shutdown()` is called no more than once.

Reply:
Done.